CorsixTH/Src/th_gfx.cpp

```diff
@@ -1033,88 +1092,119 @@ inline void THChunkRenderer::_incrementPosition(int npixels)
-                chunkFillToEndOfLine(0xFF);
-            }
-            else if(b < 0x40)
-            {
-                int amt = b;
-                if(datalen < amt)
-                    amt = datalen;
-                chunkCopy(amt, data);
-                data += amt;
-                datalen -= amt;
-            }
-            else if((b & 0xC0) == 0x80)
-            {
-                chunkFill(b - 0x80, 0xFF);
-            }
-            else
-            {
-                int amt;
-                uint8_t colour = 0;
-                if(b == 0xFF)
-                {
-                    if(datalen < 2)
-                    {
-                        break;
-                    }
-                    amt = (int)data[0];
-                    colour = data[1];
-                    data += 2;
-                    datalen -= 2;
-                }
-                else
-                {
-                    amt = b - 60 - (b & 0x80) / 2;
-                    if(datalen > 0)
-                    {
-                        colour = *data;
-                        ++data;
-                        --datalen;
-                    }
-                }
-                chunkFill(amt, colour);
-            }
+            // If the value is 0 fill to EOL with all bits set
+            chunkFillToEndOfLine(0xFF);
+        }
+        else if (inputVal < 128)
+        {
+            // The value represents the length of the buffer to copy
+            size_t amt = inputVal;
+            if (dataLen < amt)
+                amt = dataLen;
+            chunkCopy(amt, inputData);
+            inputData += amt;
+            dataLen -= amt;
+        }
+        else
+        {
+            // Fill the next (256 - value) pixels with all bits set
+            // as the value is greater than 128 because of the above check it will
+            // be the next 1-128 pixels
+            chunkFill(256 - inputVal, 0xFF);
         }
     }
-    else
+}
+
+void THChunkRenderer::decodeChunksComplex(const uint8_t * inputData, size_t dataLen)
+{
+    while (!isEndOfBuffer() && dataLen > 0)
     {
-        while(!_isDone() && datalen > 0)
+        uint8_t inputVal = *inputData;
+        --dataLen;
+        ++inputData;
+
+        if (inputVal == 0)
         {
-            uint8_t b = *data;
-            --datalen;
-            ++data;
-            if(b == 0)
-            {
-                chunkFillToEndOfLine(0xFF);
-            }
-            else if(b < 0x80)
+            // If the value is 0 fill to the end of the line with all bits set
+            chunkFillToEndOfLine(0xFF);
+        }
+        else if (inputVal < 64)
+        {
+            // Copy the number of pixels represented by inputVal.
+            size_t amt = inputVal;
+            if (dataLen < amt)
+                amt = dataLen;
+            chunkCopy(amt, inputData);
+            inputData += amt;
+            dataLen -= amt;
+        }
+        // If the input has bit 128 set without bit 64
```

Comment:
bit 128 is not 0x80, bit 128 is 2**128, which is 340282366920938463463374607431768211456 or 0x100000000000000000000000000000000 (see how hex makes bits much more clear here?).

You mean bit 7 and bit 6 here (bit b has value 2**b, starting with b=0, with value 1.